src/folding/circuits/utils.rs

```diff
@@ -1,7 +1,46 @@
 use ark_ff::PrimeField;
-use ark_r1cs_std::fields::{fp::FpVar, FieldVar};
-use ark_relations::r1cs::SynthesisError;
-use std::marker::PhantomData;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    fields::{fp::FpVar, FieldVar},
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use std::{borrow::Borrow, marker::PhantomData};
+
+/// `VecFpVar` is a wrapper around a vector of `FpVar`s.
+/// It implements the `IntoIterator` trait and can hence be iterated over
+pub struct VecFpVar<F: PrimeField> {
+    pub vec: Vec<FpVar<F>>,
+}
+
+impl<F: PrimeField> IntoIterator for VecFpVar<F> {
+    type Item = FpVar<F>;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vec.into_iter()
+    }
+}
+
+impl<F: PrimeField> AllocVar<Vec<F>, F> for VecFpVar<F> {
+    fn new_variable<T: Borrow<Vec<F>>>(
+        cs: impl Into<Namespace<F>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let cs = cs.into();
+        f().and_then(|c| {
+            let vec = c
+                .borrow()
+                .iter()
+                .map(|element| {
+                    FpVar::new_variable(cs.clone(), || Ok(element), mode)
+                        .map_err(|_| SynthesisError::Unsatisfiable)
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            Ok(Self { vec })
+        })
+    }
+}
```

Comment:
I'm not sure to understand the reason to add the `VecFpVar` type wrapper, could we not just directly use `Vec<FpVar<F>>` when needed?
In this way we can avoid the lines 9 to 43 (getting rid of the `VecFpVar` wrapper and methods)

Reply:
Later on, we will want to compute the product and also the sum of elements of a vector of `FpVar`- namely for the second part of the final sum. This abstraction helps us then do to things like:

```rust
for c_i in c {
     // ...
     let pi_thetaj = thetas[i].product()?;
     //. ...
}
// [...]
let sum = VecFpVar::new(result_vec).sum()?;
```

Also vectors of `FpVar` should be relatively common in other schemes? I think this abstraction might be helpful down the road.

That was my rationale for doing this. I might be wrong though, so open to other ways of doing things. 

What do you think?

Reply:
I'm not sure, the same logic could be said for doing type wrappers for `Vec<Fr>` (in native rust, outside circuits), but we don't add a wrapper there.

The two examples from the previous comment (sum & prod) can be done without type wrappers in one line just with things like `.iter()`:

```rust
    #[test]
    fn test_iter() {
        let v = vec![
            Fr::from(1u32),
            Fr::from(2u32),
            Fr::from(3u32),
            Fr::from(4u32),
        ];

        // in rust natively over Vec<Fr>
        let sum: Fr = v.iter().sum();
        let prod: Fr = v.iter().product();

        let cs = ConstraintSystem::<Fr>::new_ref();
        let vVar = Vec::<FpVar<Fr>>::new_witness(cs.clone(), || Ok(v.clone())).unwrap();

        // inside circuit over Vec<FpVar<Fr>>
        let sumVar: FpVar<Fr> = vVar.iter().sum();
        let prodVar: FpVar<Fr> = vVar.iter().fold(FpVar::one(), |acc, e| acc * e);

        assert_eq!(vVar.value().unwrap(), v);
        assert_eq!(sumVar.value().unwrap(), sum);
        assert_eq!(prodVar.value().unwrap(), prod);
    }
```

Reply:
Thanks, I was not aware that you could call `Vec::<FpVar<Fr>>::new_witness`! Really cool. Removed the `VecFpVar` struct [here](https://github.com/privacy-scaling-explorations/folding-schemes/pull/53/commits/5c48096c995dbd07ffe44a9b6a36f16c4e7e5c28)